spinalcordtoolbox/reports/slice.py

```diff
@@ -57,8 +57,15 @@ def __init__(self, images, p_resample=0.6):
             self._absolute_paths.append(img.absolutepath)  # change_orientation removes the field absolute_path
             img.change_orientation('SAL')
             if p_resample:
-                if i == len(images) - 1 and img.dim[3] == 1:
-                    # Last volume, if it is 3d, corresponds to a segmentation, therefore use linear interpolation here
+                # Check if image is a segmentation (binary or soft) by making sure:
+                # - 0/1 are the two most common voxel values
+                # - 0/1 account for >99% of voxels (to allow for some soft voxels)
+                unique, counts = np.unique(img.data, return_counts=True)
+                unique, counts = unique[np.argsort(counts)[::-1]], counts[np.argsort(counts)[::-1]]  # Sort by counts
+                binary_most_common = set(unique[0:2].astype(float)) == {0.0, 1.0}
+                binary_percentage = ((counts[0] + counts[1]) / np.sum(counts))
+                if binary_most_common and binary_percentage > 0.99:
```

Comment:
I'm not 100% sure this is a foolproof method for detecting segmentations. Is there any chance a non-segmentation would be 99% 0/1 voxels? Or, is there any chance that a softseg would have <99% 0/1 voxels?

Reply:
Yeah, this is tricky. It really depends on the character of the lesion (how it was segmented, how the rater considered the partial volume effect).
I didn't calculate the voxel intensity values distribution for soft seg, though. Maybe we could decrease it to 95% to be less conservative?